packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -74,324 +186,476 @@ class FocusNode extends ChangeNotifier {
+    final List<DiagnosticsNode> children = <DiagnosticsNode>[];
+    int count = 1;
+    for (FocusNode child in _children) {
+      children.add(child.toDiagnosticsNode(name: 'child $count'));
+      ++count;
+    }
+    return children;
+  }
+}
+
+/// A [FocusNode] subclass that acts as a scope for its descendants,
+/// maintaining state about which descendant was last or is currently focused.
+class FocusScopeNode extends FocusNode {
+  /// Creates a FocusScope node.
+  ///
+  /// All parameters are optional.
+  FocusScopeNode({
+    BuildContext context,
+    String debugLabel,
+    FocusOnKeyCallback onKey,
+  }) : super(context: context, debugLabel: debugLabel, onKey: onKey);
+
+  @override
+  FocusScopeNode get nearestScope => this;
+
+  @override
+  void reparentIfNeeded(BuildContext childContext, FocusNode child) {
+    final FocusScopeNode previousEnclosingScope = child.enclosingScope;
+    final bool hadChildren = _children.isNotEmpty;
+    super.reparentIfNeeded(childContext, child);
+    final FocusScopeNode currentEnclosingScope = child.enclosingScope;
+    // If the child moved scopes, then the policy data is invalid.
+    if (previousEnclosingScope != currentEnclosingScope) {
+      policyData = null;
+    }
+    // If we just added our first child to this scope, and this scope had the
+    // focus, then focus the child, otherwise make it the current focus.
+    if (!hadChildren && currentEnclosingScope.focusedChild == null && currentEnclosingScope.hasFocus) {
+      child.requestFocus();
+    }
+  }
+
+  /// Returns true if this scope is the focused child of its parent scope.
+  bool get isFirstFocus => enclosingScope.focusedChild == this;
+
+  /// Add the given node as a child of this focus scope, and make it be the
+  /// [focusedChild].
+  void setFirstFocus(BuildContext childContext, FocusScopeNode node) {
     assert(node != null);
-    if (_focus != node)
-      return;
-    _focus._parent = null;
-    _focus._manager = null;
-    _focus = null;
-    _didChangeFocusChain();
+    if (node._parent == null) {
+      reparentIfNeeded(childContext, node);
+    }
+    assert(node.ancestors.contains(this));
+    node.requestFocus();
   }
 
-  /// Makes the given child the first focus of this scope.
+  /// If this scope lacks a focus, reparent the node as a child of this one, and
+  /// request that the given node becomes the focus.
   ///
-  /// If the child has another parent scope, the child is first removed from
-  /// that scope. After this method returns [isFirstFocus] will be true for
-  /// the child.
-  void setFirstFocus(FocusScopeNode child) {
-    assert(child != null);
-    if (_firstChild == child)
-      return;
-    child.detach();
-    _prepend(child);
-    assert(child._parent == this);
-    _didChangeFocusChain();
+  /// Useful for widgets that wish to grab the focus if no other widget already
+  /// has the focus.
+  ///
+  /// The node is notified that it has received the overall focus in a
```

Comment:
It's not obvious from just looking at this method's implementation that that's what's going on. If there really is a microtask involved, it would be helpful to explain what the implications are.